src/backend/brokers/go.emails/encrypt.go

```diff
@@ -0,0 +1,148 @@
+// Copyleft (ɔ) 2017 The Caliopen contributors.
+// Use of this source code is governed by a GNU AFFERO GENERAL PUBLIC
+// license (AGPL) that can be found in the LICENSE file.
+
+// package email_broker handles codec/decodec between external emails and Caliopen message format
+package email_broker
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/base64"
+	. "github.com/CaliOpen/Caliopen/src/backend/defs/go-objects"
+	log "github.com/Sirupsen/logrus"
+	"github.com/emersion/go-message"
+	"math/rand"
+	"net/mail"
+	"strings"
+	"time"
+)
+
+var boundaryChars = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789+:=?")
+
+func RandomString(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = boundaryChars[rand.Intn(len(boundaryChars))]
+	}
+	return string(b)
+}
+
+// Return a MIME boudary as specified in RFC 1341 (7.2.1)
+//
+// boundary := 0*69<bchars> bcharsnospace
+//
+// bchars := bcharsnospace / " "
+//
+// bcharsnospace :=    DIGIT / ALPHA / "'" / "(" / ")" / "+"  /
+// "_"
+//                / "," / "-" / "." / "/" / ":" / "=" / "?"
+func (b *EmailBroker) NewBoundary() string {
+	return RandomString(42)
+}
+
+func (b *EmailBroker) NewMessageId(uuid []byte) string {
+	// sha256 internal message id to form external message id
+	hasher := sha256.New()
+	hasher.Write(uuid)
+	sha := base64.URLEncoding.EncodeToString(hasher.Sum(nil))
+	messageId := sha + "@" + b.Config.PrimaryMailHost // should be the default domain in case there are multiple 'from' addresses
+	return messageId
+}
+
+//Marshal an encrypted email
+func (b *EmailBroker) MarshalEncryptedEmail(msg *Message, em *EmailMessage) (err error) {
+
+	mainHeader := make(message.Header)
+	params := map[string]string{"boundary": b.NewBoundary(), "protocol": "application/pgp-encrypted"}
+	mainHeader.Set("Subject", msg.Subject)
+	mainHeader.Set("Date", time.Now().Format(time.RFC1123Z))
+	mainHeader.SetContentType("multipart/encrypted", params)
+	mainHeader.Set("X-Mailer", "Caliopen-"+b.Config.AppVersion)
+
+	addr_fields := formatAddressList(msg)
+	for field, addrs := range addr_fields {
+		if len(addrs) > 0 {
+			mainHeader.Set(field, strings.Join(addrs, ","))
+		}
+	}
+
+	// References headers
+	mainHeader.Set("Message-ID", "<"+b.NewMessageId(em.Message.Message_id.Bytes())+">")
+
+	// Formatted multipart body
+	body, err := formatBody(msg, mainHeader)
+	if err != nil {
+		return err
+	}
+	em.Email.Raw = body
```

Comment:
this will be overwritten by MarshalEmail !

Reply:
nope I shunt the MarshalEmail process if it's encrypted to produce a full mail.